File: app/src/main/java/com/example/coviddefender/RecyclerViewAdapter/AnnouncementAdapter.java
package com.example.coviddefender.RecyclerViewAdapter;

import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.coviddefender.R;
import com.example.coviddefender.entity.Announcement;
import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
import com.firebase.ui.firestore.FirestoreRecyclerOptions;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.card.MaterialCardView;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.squareup.picasso.Picasso;

public class AnnouncementAdapter extends FirestoreRecyclerAdapter<Announcement, AnnouncementAdapter.AnnouncementViewHolder> {
    FirebaseStorage storage;
    StorageReference storageReference;

    public AnnouncementAdapter(@NonNull FirestoreRecyclerOptions<Announcement> options) {
        super(options);
    }

    @Override
    protected void onBindViewHolder(@NonNull AnnouncementAdapter.AnnouncementViewHolder holder, int position, @NonNull Announcement model) {
        holder.tv_description.setText(model.getDescription());
        Uri url = Uri.parse(model.getUrl());
        holder.announcement_card.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // open browser with link
                Intent intent = new Intent(Intent.ACTION_VIEW, url);
                view.getContext().startActivity(Intent.createChooser(intent, "Browse with"));
            }
        });

        // set up image view
        storage = FirebaseStorage.getInstance();
        storageReference = storage.getReference();
        storageReference.child(model.getImage()).getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
            @Override
            public void onSuccess(Uri uri) {
                Picasso.get().load(uri).into(holder.announcement_image);
            }
        });
    }

    @NonNull
    @Override
    public AnnouncementAdapter.AnnouncementViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.latest_announcement_item, parent, false);
        return new AnnouncementAdapter.AnnouncementViewHolder(view);
    }


    public class AnnouncementViewHolder extends RecyclerView.ViewHolder {
        private final TextView tv_description;
        private final MaterialCardView announcement_card;
        private final ImageView announcement_image;

        public AnnouncementViewHolder(View view) {
            super(view);
            tv_description = view.findViewById(R.id.announcement_text);
            announcement_card = view.findViewById(R.id.announcement_card);
            announcement_image = view.findViewById(R.id.announcement_image);
        }
    }
}

File: app/src/main/java/com/example/coviddefender/RecyclerViewAdapter/AnnouncementAdapter.kt
package com.example.coviddefender.RecyclerViewAdapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.coviddefender.R
import com.example.coviddefender.entity.Announcement

class AnnouncementAdapter(var announcements: ArrayList<Announcement>) :
    RecyclerView.Adapter<AnnouncementAdapter.AnnouncementViewHolder>()  // interface
{
    // Second Called
    // Constructor for viewholder
    // provide reference to the type of views that you are using
    class AnnouncementViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        val thumbnail: ImageView = itemView.findViewById<ImageView>(R.id.announcement_image)
        val tv_description: TextView = itemView.findViewById<TextView>(R.id.announcement_text)
        fun bind(announcement: Announcement) {
            tv_description.text = announcement.description
            thumbnail.setImageResource(announcement.thumbnail)
        }
    }

    // First function called
    // create new views (invoked by layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AnnouncementViewHolder {
        // create new view, defines the UI of the list item
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.latest_announcement_item,parent,false)

        return AnnouncementViewHolder(view)
    }

    // Third called
    // replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: AnnouncementViewHolder, position: Int) {
        holder.bind(announcements[position])
    }

    override fun getItemCount(): Int = announcements.size
}